keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1550 @@
++    // Key is the resource type and values is the class name within the resource type.
++    map<string, string> classes = 1;
++}
+```
+
+The `UpdateContainerResourcesRequest` message will be similarly extended to
+allow updating of QoS-class resource configuration of a running container.
+Depending on runtime-level support of a particular resource (and possibly the
+type of resource) UpdateContainerResourcesRequest might fail. Later phases
+(with resource discovery/status) adds the ability to distinguish immutable
+resource types. Note that neither of the existing QoS-class resource types (RDT
+or blockio) support updates because of runtime limitations, yet.
+
+```diff
+ message UpdateContainerResourcesRequest {
+
+ ...
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++    // Configuration of class resources.
++    ContainerClassResources class_resources = 5;
+}
+```
+
+The `PodSandboxConfig` will be supplemented with a corresponding
+`class_resources` field that will be the Pod level configuration. Depending on
+the resource this might be interpreted as a pod-level default (that is used if
+nothing is specified in the `ContainerConfig`) or as a true Pod-level setting -
+in the end the detailed behavior will be responsibility of the container
+runtime.
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS-class resources.
++    PodClassResources class_resources = 10;
++
+ }
+
++// PodClassResources specifies the configuration of QoS-class resources
++// resources of a pod.
++message PodClassResources {
++    // QoS-class resource assignment of the pod.
++    // Key is the resource type and values is the class name within the resource type.
++    map<string, string> class = 1;
++}
+```
+
+Also, define "known" QoS-class resource types to more easily align container
+runtime implementations:
+
+```diff
++
++const (
++       // ClassResourceRdt is the name of the RDT QoS-class resource
++       ClassResourceRdt = "rdt"
++       // ClassResourceBlockio is the name of the blockio QoS-class resource
++       ClassResourceBlockio = "blockio"
++)
+```
+
+### Pod annotations
+
+Use Pod annotation as the initial K8s user interface, similar to e.g. how
+seccomp support was added. This will bridge the gap between enabling
+QoS-class resources in the CRI protocol and making them available in the Pod
+spec.
+
+A feature gate ClassResources enables kubelet to look for pod
+annotations and set the QoS-class resource assignment via CRI protocol accordingly.
+
+Specifically, kubelet will support annotations for specifying RDT and blockio
+class, the two types of QoS-class resources that already have basic support in the
+container runtimes.
+
+  class for all containers
```

Comment:
I'm confused .above you mention we'll change the cri message to include a Resources field, but down here you're saying we'll use annotations?

Reply:
Above, we introduce these new fields in the CRI API but PodSpec is still missing. Pod annotations interpreted by kubelet would serve as the "phase 1 user interface".

Reply:
no annotations anymore in the proposal -> "resolved"